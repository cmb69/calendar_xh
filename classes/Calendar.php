<?php

/**
 * Copyright 2021 Christoph M. Becker
 *
 * This file is part of Calendar_XH.
 *
 * Calendar_XH is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Calendar_XH is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Calendar_XH.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace Calendar;

class Calendar
{
    /** @var bool */
    private $weekStartsMonday;

    /**
     * @param bool $weekStartsMonday
     */
    public function __construct($weekStartsMonday)
    {
        $this->weekStartsMonday = $weekStartsMonday;
    }

    /**
     * @param int $year
     * @param int $month
     * @return array<int,array<int,(int|null)>>
     */
    public function getMonthMatrix($year, $month)
    {
        $result = [];
        $row = [];
        $days = (int) date('t', mktime(1, 1, 1, $month, 1, $year));
        for ($day = 1; $day <= $days; $day++) {
            $dayofweek = $this->getDayOfWeek($year, $month, $day);
            if ($day === 1) {
                for ($i = 0; $i < $dayofweek; $i++) {
                    $row[] = null;
                }
            }
            $row[] = $day;
            if ($day === $days) {
                for ($i = $dayofweek + 1; $i < 7; $i++) {
                    $row[] = null;
                }
            }
            if (count($row) === 7 || $day === $days) {
                $result[] = $row;
                $row = [];
            }
        }
        return $result;
    }

    /**
     * @param int $year
     * @param int $month
     * @param int $day
     * @return int
     */
    private function getDayOfWeek($year, $month, $day)
    {
        $dayofweek = (int) date('w', mktime(1, 1, 1, $month, $day, $year));
        if ($this->weekStartsMonday) {
            $dayofweek = ($dayofweek + 6) % 7;
        }
        return $dayofweek;
    }
}
