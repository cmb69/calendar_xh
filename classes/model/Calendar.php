<?php

/**
 * Copyright (c) Christoph M. Becker
 *
 * This file is part of Calendar_XH.
 *
 * Calendar_XH is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Calendar_XH is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Calendar_XH.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace Calendar\Model;

use Calendar\Html2Text;

class Calendar
{
    /** @var array<string,Event> */
    private $events;

    /** @param array<string,Event> $events */
    public function __construct(array $events)
    {
        $this->events = $events;
    }

    /** @return array<string,Event> */
    public function events(): array
    {
        return $this->events;
    }

    public function event(string $id): ?Event
    {
        return $this->events[$id] ?? null;
    }

    public function nextEvent(LocalDateTime $now): ?Event
    {
        $nextevent = null;
        $nextldt = null;
        foreach ($this->events as $event) {
            $ldt = $event->after($now);
            if ($ldt === null) {
                continue;
            }
            if ($nextldt === null || $ldt->compare($nextldt) < 0) {
                $nextevent = $event;
                $nextldt = $ldt;
            }
        }
        return $nextevent;
    }

    /** @return list<Event> */
    public function eventsDuring(int $year, int $month): array
    {
        $result = [];
        foreach ($this->events as $event) {
            if ($event->occursDuring($year, $month)) {
                $result[] = $event;
            }
        }
        uasort($result, function (Event $a, Event $b) use ($year): int {
            $dt1 = $a->isBirthday() ? $a->start()->withYear($year) : $a->start();
            $dt2 = $b->isBirthday() ? $b->start()->withYear($year) : $b->start();
            return $dt1->compare($dt2);
        });
        return $result;
    }

    /** @return list<Event> */
    public function eventsOn(LocalDateTime $day, bool $daysBetween): array
    {
        assert($day->hour() === 0 && $day->minute() === 0);
        $result = [];
        foreach ($this->events as $event) {
            if ($event->occursOn($day, $daysBetween)) {
                $result[] = $event;
            }
        }
        return $result;
    }

    public function delete(string $id): void
    {
        assert(array_key_exists($id, $this->events));
        unset($this->events[$id]);
    }

    public function toICalendarString(Html2Text $converter, string $host): string
    {
        $res = "BEGIN:VCALENDAR\r\n"
            . "PRODID:-//3-magi.net//Calendar_XH//EN\r\n"
            . "VERSION:2.0\r\n";
        foreach ($this->events() as $id => $event) {
            $res .= $event->toICalendarString($id, $converter, $host);
        }
        $res .= "END:VCALENDAR\r\n";
        return $res;
    }
}
