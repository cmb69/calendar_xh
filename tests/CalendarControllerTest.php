<?php

/**
 * Copyright 2023 Christoph M. Becker
 *
 * This file is part of Calendar_XH.
 *
 * Calendar_XH is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Calendar_XH is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Calendar_XH.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace Calendar;

use ApprovalTests\Approvals;
use Calendar\Infra\Counter;
use Calendar\Infra\DateTimeFormatter;
use Calendar\Model\Calendar;
use Calendar\Model\Event;
use Calendar\Model\LocalDateTime;
use Calendar\Model\NoRecurrence;
use Calendar\Model\YearlyRecurrence;
use org\bovigo\vfs\vfsStream;
use PHPUnit\Framework\TestCase;
use Plib\DocumentStore;
use Plib\FakeRequest;
use Plib\View;

class CalendarControllerTest extends TestCase
{
    /** @var array<string,string> */
    private $conf;

    /** @var array<string,string */
    private $lang;

    /** @var DocumentStore */
    private $store;

    /** @var DateTimeFormatter */
    private $dateTimeFormatter;

    /** @var Counter */
    private $counter;

    /** @var View */
    private $view;

    public function setUp(): void
    {
        vfsStream::setup();
        $this->conf = XH_includeVar("./config/config.php", "plugin_cf")["calendar"];
        $this->lang = XH_includeVar("./languages/en.php", "plugin_tx")["calendar"];
        $this->store = new DocumentStore(vfsStream::url("root/"));
        $calendar = Calendar::updateIn($this->store);
        $calendar->addEvent("111", $this->lunchBreak()->toDto());
        $calendar->addEvent("222", $this->weekend()->toDto());
        $calendar->addEvent("333", $this->birthday()->toDto());
        $this->store->commit();
        $this->dateTimeFormatter = new DateTimeFormatter($this->lang);
        $this->view = new View("./views/", $this->lang);
        $this->counter = new Counter(1);
    }

    private function sut(): CalendarController
    {
        return new CalendarController(
            "./",
            $this->conf,
            $this->store,
            $this->dateTimeFormatter,
            1,
            $this->counter,
            $this->view
        );
    }

    public function testDefaultActionRendersHtml()
    {
        $request = new FakeRequest([
            "url" => "http://example.com/?page",
            "time" => strtotime("2023-01-30T14:27:00+00:00"),
        ]);
        $response = $this->sut()(0, 0, "", false, $request);
        Approvals::verifyHtml($response->output());
    }

    /** @dataProvider monthData */
    public function testShowsMonthAsRequested(int $month, int $year, string $url, string $expected): void
    {
        $request = new FakeRequest(["url" => $url, "time" => strtotime("2023-01-30T14:27:00+00:00")]);
        $response = $this->sut()($year, $month, "", false, $request);
        $this->assertStringContainsString($expected, $response->output());
    }

    public function monthData(): array
    {
        return [
            [
                0, 0,
                "http://example.com/?Events",
                "January 2023",
            ],
            [
                0, 2020,
                "http://example.com/?Events",
                "January 2020",
            ],
            [
                3, 2020,
                "http://example.com/?Events",
                "March 2020",
            ],
            [
                0, 0,
                "http://example.com/?Events&year=2021&month=7",
                "July 2021",
            ],
            [
                8, 2020,
                "http://example.com/?Events&year=2021&month=7",
                "August 2020",
            ],
            [
                13, -1,
                "http://example.com/?Events&year=2021&month=7",
                "December 0001",
            ],
        ];
    }

    public function testFailsToRenderBigCalendarIfEventPagesAreDisabled(): void
    {
        $request = new FakeRequest([
            "url" => "http://example.com/?page",
            "time" => strtotime("2023-01-30T14:27:00+00:00"),
        ]);
        $response = $this->sut()(0, 0, "", true, $request);
        $this->assertStringContainsString("Big calendars require event pages to be enabled!", $response->output());
    }

    public function testRendersBigCalendar(): void
    {
        $this->conf["event_allow_single"] = "true";
        $request = new FakeRequest([
            "url" => "http://example.com/?page",
            "time" => strtotime("2023-01-30T14:27:00+00:00"),
        ]);
        $response = $this->sut()(0, 0, "", true, $request);
        Approvals::verifyHtml($response->output());
    }

    private function lunchBreak(): Event
    {
        $start = new LocalDateTime(2023, 1, 4, 12, 0);
        $end = new LocalDateTime(2023, 1, 4, 13, 0);
        $url = "http://example.com/lunchbreak";
        $recurrence = new NoRecurrence($start, $end);
        return new Event("", $start, $end, "Lunch break", $url, "Tips for lunch breaks", "whereever I am", $recurrence);
    }

    private function weekend(): Event
    {
        $start = new LocalDateTime(2023, 1, 7, 0, 0);
        $end = new LocalDateTime(2023, 1, 8, 23, 59);
        $recurrence = new NoRecurrence($start, $end);
        return new Event("", $start, $end, "Weekend", "", "", "", $recurrence);
    }

    private function birthday(): Event
    {
        $start = new LocalDateTime(2000, 1, 1, 0, 0);
        $end = new LocalDateTime(2000, 1, 1, 23, 59);
        $recurrence = new YearlyRecurrence($start, $end, null);
        return new Event("", $start, $end, "Millenium", "", "", "###", $recurrence);
    }
}
