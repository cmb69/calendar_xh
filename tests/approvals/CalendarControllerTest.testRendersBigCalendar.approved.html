
<table class="calendar_main calendar_big" role="presentation">
  <caption class="calendar_monthyear">
    <a href="/?page&amp;month=12&amp;year=2022" aria-labelledby="calendar_id_1">
      <span>◄</span>
      <span role="tooltip" id="calendar_id_1">Previous month</span>
    </a>
    <span>January 2023</span>
    <a href="/?page&amp;month=2&amp;year=2023" aria-labelledby="calendar_id_2">
      <span>►</span>
      <span role="tooltip" id="calendar_id_2">Next month</span>
    </a>
  </caption>
  <tr>
    <th class="calendar_daynames calendar_day" scope="col">Monday</th>
    <th class="calendar_daynames calendar_day" scope="col">Tuesday</th>
    <th class="calendar_daynames calendar_day" scope="col">Wednesday</th>
    <th class="calendar_daynames calendar_day" scope="col">Thursday</th>
    <th class="calendar_daynames calendar_day" scope="col">Friday</th>
    <th class="calendar_daynames calendar_we" scope="col">Saturday</th>
    <th class="calendar_daynames calendar_we" scope="col">Sunday</th>
  <tr>
  <tr>
    <td class="calendar_noday">
      <div>
        <span></span>
      </div>
    </td>
    <td class="calendar_noday">
      <div>
        <span></span>
      </div>
    </td>
    <td class="calendar_noday">
      <div>
        <span></span>
      </div>
    </td>
    <td class="calendar_noday">
      <div>
        <span></span>
      </div>
    </td>
    <td class="calendar_noday">
      <div>
        <span></span>
      </div>
    </td>
    <td class="calendar_noday">
      <div>
        <span></span>
      </div>
    </td>
    <td class="calendar_we">
      <div>
        <span>1</span>
        <a href="http://example.com/?&amp;function=calendar_event&amp;event_id=333&amp;calendar_occurrence=2023-01-01">Millenium</a>
      </div>
    </td>
  </tr>
  <tr>
    <td class="calendar_day">
      <div>
        <span>2</span>
      </div>
    </td>
    <td class="calendar_day">
      <div>
        <span>3</span>
      </div>
    </td>
    <td class="calendar_day">
      <div>
        <span>4</span>
        <a href="http://example.com/lunchbreak">Lunch break</a>
      </div>
    </td>
    <td class="calendar_day">
      <div>
        <span>5</span>
      </div>
    </td>
    <td class="calendar_day">
      <div>
        <span>6</span>
      </div>
    </td>
    <td class="calendar_we">
      <div>
        <span>7</span>
        <a href="http://example.com/?&amp;function=calendar_event&amp;event_id=222&amp;calendar_occurrence=2023-01-07">Weekend</a>
      </div>
    </td>
    <td class="calendar_we">
      <div>
        <span>8</span>
        <a href="http://example.com/?&amp;function=calendar_event&amp;event_id=222&amp;calendar_occurrence=2023-01-07">Weekend</a>
      </div>
    </td>
  </tr>
  <tr>
    <td class="calendar_day">
      <div>
        <span>9</span>
      </div>
    </td>
    <td class="calendar_day">
      <div>
        <span>10</span>
      </div>
    </td>
    <td class="calendar_day">
      <div>
        <span>11</span>
      </div>
    </td>
    <td class="calendar_day">
      <div>
        <span>12</span>
      </div>
    </td>
    <td class="calendar_day">
      <div>
        <span>13</span>
      </div>
    </td>
    <td class="calendar_we">
      <div>
        <span>14</span>
      </div>
    </td>
    <td class="calendar_we">
      <div>
        <span>15</span>
      </div>
    </td>
  </tr>
  <tr>
    <td class="calendar_day">
      <div>
        <span>16</span>
      </div>
    </td>
    <td class="calendar_day">
      <div>
        <span>17</span>
      </div>
    </td>
    <td class="calendar_day">
      <div>
        <span>18</span>
      </div>
    </td>
    <td class="calendar_day">
      <div>
        <span>19</span>
      </div>
    </td>
    <td class="calendar_day">
      <div>
        <span>20</span>
      </div>
    </td>
    <td class="calendar_we">
      <div>
        <span>21</span>
      </div>
    </td>
    <td class="calendar_we">
      <div>
        <span>22</span>
      </div>
    </td>
  </tr>
  <tr>
    <td class="calendar_day">
      <div>
        <span>23</span>
      </div>
    </td>
    <td class="calendar_day">
      <div>
        <span>24</span>
      </div>
    </td>
    <td class="calendar_day">
      <div>
        <span>25</span>
      </div>
    </td>
    <td class="calendar_day">
      <div>
        <span>26</span>
      </div>
    </td>
    <td class="calendar_day">
      <div>
        <span>27</span>
      </div>
    </td>
    <td class="calendar_we">
      <div>
        <span>28</span>
      </div>
    </td>
    <td class="calendar_we">
      <div>
        <span>29</span>
      </div>
    </td>
  </tr>
  <tr>
    <td class="calendar_today calendar_day">
      <div>
        <span>30</span>
      </div>
    </td>
    <td class="calendar_day">
      <div>
        <span>31</span>
      </div>
    </td>
    <td class="calendar_noday">
      <div>
        <span></span>
      </div>
    </td>
    <td class="calendar_noday">
      <div>
        <span></span>
      </div>
    </td>
    <td class="calendar_noday">
      <div>
        <span></span>
      </div>
    </td>
    <td class="calendar_noday">
      <div>
        <span></span>
      </div>
    </td>
    <td class="calendar_noday">
      <div>
        <span></span>
      </div>
    </td>
  </tr>
</table>
